### PR TITLE
[Admin] Add type hints to all files in the Admin dir

### PR DESCRIPTION
**Pull request by tambait**
## Subject

This is a follow up of #6404 to add param & return type hints.

## Changelog

```markdown
### Added
- Added type hints to all files in the Admin dir.

### CODE REVIEW

File: src/Admin/AdminExtensionInterface.php [hunk 1]
```diff
@@ -155,13 +161,15 @@ public function configureActionButtons(
     /**
      * Returns a list of default filters.
      *
-     * @phpstan-param AdminInterface<object> $admin
```
Comment (VincentLanglet) [inline src/Admin/AdminExtensionInterface.php:158]: should be kept

File: src/Admin/FieldDescriptionInterface.php [hunk 2]
```diff
@@ -18,36 +18,23 @@
...
-     * Set the name.
-     */
     public function setName(?string $name): void;
 
     /**
      * Returns the name, the name can be used as a form label or table header.
-     *
-     * @return string the name
      */
     public function getName(): ?string;
```
Comment (VincentLanglet) [inline src/Admin/FieldDescriptionInterface.php:30]: Will be `:string` after the 3.X merge into master

File: src/Admin/FieldDescriptionInterface.php [hunk 3]
```diff
@@ -18,36 +18,23 @@
...
-     * Returns the field name.
-     *
-     * @return string the field name
-     */
     public function getFieldName(): ?string;
 
-    /**
-     * Set the name.
-     */
     public function setName(?string $name): void;
```
Comment (VincentLanglet) [inline src/Admin/FieldDescriptionInterface.php:25]: Should be `string $name`

File: src/Admin/AbstractAdmin.php [hunk 4]
```diff
@@ -2073,7 +2090,7 @@ public function getRouteBuilder(): ?RouteBuilderInterface
         return $this->routeBuilder;
     }
 
-    public function toString(object $object): string
+    public function toString(?object $object): string
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2093]:
```suggestion
    public function toString(object $object): string
```

ToString should not allow null as param.

File: src/Admin/AbstractAdmin.php [hunk 5]
```diff
@@ -2580,7 +2597,7 @@ protected function buildForm(): void
     /**
      * Gets the subclass corresponding to the given name.
      *
-     * @phpstan-return class-string<T>
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2583]: This seems weird

File: src/Admin/AbstractAdmin.php [hunk 6]
```diff
@@ -2676,16 +2693,15 @@ protected function configureDefaultFilterValues(array &$filterValues): void
...
      */
-    protected function configureDefaultSortValues(array &$sortValues)
+    protected function configureDefaultSortValues(array &$sortValues): void
     {
     }
 
     /**
      * Set the parent object, if any, to the provided object.
-     *
-     * @phpstan-param T $object
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2688]: These removal seems weird

File: src/Admin/AdminInterface.php [hunk 7]
```diff
@@ -102,11 +104,12 @@ public function setTemplates(array $templates): void;
      */
     public function setTemplate(string $name, string $template): void;
 
+    public function setModelManager(ModelManagerInterface $modelManager): void;
```
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:107]: Why is this needed ?

File: src/Admin/AdminInterface.php [hunk 8]
```diff
@@ -102,11 +104,12 @@ public function setTemplates(array $templates): void;
...
+
     public function getModelManager(): ?ModelManagerInterface;
 
-    /**
-     * @return string the manager type of the admin
-     */
+    public function setManagerType(string $managerType): void;
```
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:111]: Why is this needed ?

File: src/Admin/AdminInterface.php [hunk 9]
```diff
@@ -231,7 +236,7 @@ public function getRouteBuilder(): ?RouteBuilderInterface;
     /**
      * @phpstan-param T $object
      */
-    public function toString(object $object): string;
+    public function toString(?object $object): string;
```
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:239]: null shouldnt be allowed

File: src/Admin/BreadcrumbsBuilder.php [hunk 10]
```diff
@@ -70,7 +70,7 @@ public function getBreadcrumbs(AdminInterface $admin, string $action): array
      *
      * Note: the method will be called by the top admin instance (parent => child)
      */
-    public function buildBreadcrumbs(
+    private function buildBreadcrumbs(
```
Comment (VincentLanglet) [inline src/Admin/BreadcrumbsBuilder.php:73]: why this change ? This need a deprecation in 3.x.

File: src/Admin/AdminInterface.php (hunk 7 above)
Comment (tambait) [reply src/Admin/AdminInterface.php:107]: Not sure tbh :), I saw `setModelManager` and 'setManagerType` additions in @jordisala1991  PR

File: src/Admin/AbstractAdmin.php [hunk 11]
```diff
@@ -2676,9 +2694,10 @@ protected function configureDefaultFilterValues(array &$filterValues): void
      *   $sortValues['_sort_by'] = 'foo'
      *   $sortValues['_sort_order'] = 'DESC'
      *
+     * @param array{_page?: int, _per_page?: int, _sort_by?: string, _sort_order?: string} $sortValues*
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2697]: The detail is already present in phpstan-param, so we can remove this.

Or use array<string, string|int> which has a better integration by Phpstorm.

File: src/Admin/AdminExtensionInterface.php [hunk 12]
```diff
@@ -143,6 +145,12 @@ public function postRemove(AdminInterface $admin, object $object): void;
     /**
      * Get all action buttons for an action.
      *
+     * @param array<string, array<string, mixed>> $list
+     *
+     * @phpstan-param AdminInterface<object> $admin
```
Comment (VincentLanglet) [inline src/Admin/AdminExtensionInterface.php:150]: The phpstan param is duplicated

File: src/Admin/AdminExtensionInterface.php [hunk 13]
```diff
@@ -155,13 +163,15 @@ public function configureActionButtons(
...
      *
-     * @phpstan-param AdminInterface<object> $admin
+     * @param array<string, array<string, mixed>> $filterValues
      */
     public function configureDefaultFilterValues(AdminInterface $admin, array &$filterValues): void;
 
     /**
      * Returns a list of default sort values.
      *
+     * @param array{_page?: int, _per_page?: int, _sort_by?: string, _sort_order?: string} $sortValues
```
Comment (VincentLanglet) [inline src/Admin/AdminExtensionInterface.php:173]: Similar. It would be better to use phpstan param for this

File: src/Admin/AdminHelper.php [hunk 14]
```diff
@@ -79,6 +75,8 @@ public function getChildFormView(FormView $formView, string $elementId): ?FormVi
      * @throws \RuntimeException
      * @throws \Exception
      *
+     * @return array{0: FieldDescriptionInterface|null, 1: FormInterface}
```
Comment (VincentLanglet) [inline src/Admin/AdminHelper.php:78]: There is phpstan return for this

File: src/Admin/BaseFieldDescription.php [hunk 15]
```diff
@@ -119,7 +119,7 @@ abstract class BaseFieldDescription implements FieldDescriptionInterface
     protected $associationAdmin;
 
     /**
-     * @var array[] cached object field getters
+     * @var array<string, array{method: 'getter'|'call'|'var', getter?: string}>
```
Comment (VincentLanglet) [inline src/Admin/BaseFieldDescription.php:122]:
```suggestion
     * @var string[][]
     * @phpstan-var array<string, array{method: 'getter'|'call'|'var', getter?: string}>
```

File: src/Admin/BaseFieldDescription.php [hunk 16]
```diff
@@ -438,6 +443,9 @@ private function callCachedGetter(object $object, string $fieldName, array $para
         return $object->{$fieldName};
     }
 
+    /**
+     * @param 'call'|'getter'|'var' $method
```
Comment (VincentLanglet) [inline src/Admin/BaseFieldDescription.php:447]:
```suggestion
     * @phpstan-param 'call'|'getter'|'var' $method
```

File: src/Admin/UrlGeneratorInterface.php [hunk 17]
```diff
@@ -68,7 +68,7 @@ public function generateUrl(string $name, array $parameters = [], int $reference
      *
      * @param array<string, mixed> $parameters
      *
-     * @return array return url parts: 'route', 'routeParameters', 'routeAbsolute'
+     * @return array{route: string, routeParameters: array<string, string>, routeAbsolute: bool}
```
Comment (VincentLanglet) [inline src/Admin/UrlGeneratorInterface.php:71]:
```suggestion
     * @return array<string, mixed>
     * @phpstan-return array{route: string, routeParameters: array<string, string>, routeAbsolute: bool}
```

File: src/Admin/AdminExtensionInterface.php [hunk 18]
```diff
@@ -155,7 +161,7 @@ public function configureActionButtons(
     /**
      * Returns a list of default filters.
      *
-     * @phpstan-param AdminInterface<object> $admin
```
[the remark this answers is not in the document]
Comment (tambait) [reply src/Admin/AdminExtensionInterface.php:158]: Sorry, it seems I didn't push fixup for it

File: src/Admin/AdminHelper.php [hunk 19]
```diff
@@ -191,11 +187,10 @@ public function appendFormFieldElement(AdminInterface $admin, object $subject, s
      *
      * @param string $elementId expects string in format used in form id field.
      *                          (uniqueIdentifier_model_sub_model or uniqueIdentifier_model_1_sub_model etc.)
-     * @param mixed  $model
      *
      * @throws \Exception
      */
-    public function getElementAccessPath(string $elementId, $model): string
+    public function getElementAccessPath(string $elementId, object $model): string
```
[the remark this answers is not in the document]
Comment (tambait) [reply src/Admin/AdminHelper.php:193]: I ported that from @jordisala1991  PR, so I'm not sure if it should be mixed or or there were some changes in the code. But, looking now at `isReadable` which takes `$model` as a first parameter, it expects it to be `@param object|array`

[the remark this answers is not in the document]
Comment (VincentLanglet) [reply src/Admin/AdminHelper.php:193]: Yeah, I know you got that from jordi PR. 

I just added a comment in order to check if it's right. (I was on phone).

The only usage is
```
$this->getElementAccessPath($elementId, $subject);
```
Where `$subject` is coming from `appendFormFieldElement`

But now I think that `getElementAccessPath` should be a private method since it's only used in AdminHelper.

File: tests/Admin/AdminTest.php [hunk 20]
```diff
@@ -745,8 +745,6 @@ public function testGetLabelTranslatorStrategy(): void
     {
         $admin = new PostAdmin('sonata.post.admin.post', 'NewsBundle\Entity\Post', 'Sonata\NewsBundle\Controller\PostAdminController');
 
-        $this->assertNull($admin->getLabelTranslatorStrategy());
```
Comment (VincentLanglet) [inline tests/Admin/AdminTest.php:748]: I assume you remove this because it throws a TypeError.

IMHO, We should throw a LogicException in these getter when the value is null.
And we could then add a test with `$this->expectException(\LogicException::class)`.

But I'd like the @sonata-project/contributors point of view.
WDYT @jordisala1991 ?

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: tests/Admin/AdminTest.php (hunk 20 above)
Comment (jordisala1991) [reply tests/Admin/AdminTest.php:748]: I would be a good idea, yes. I removed it to just make the test pass, but would be better to add a test instead

File: tests/Admin/AdminHelperTest.php [hunk 21]
```diff
@@ -84,7 +84,11 @@ public function testGetElementAccessPath(): void
         $subObject->expects($this->atLeastOnce())->method('getAnother')->willReturn($sub2Object);
         $sub2Object->expects($this->atLeastOnce())->method('getMoreThings')->willReturn('Value');
 
-        $path = $this->helper->getElementAccessPath('uniquePartOfId_path_to_object_0_another_more_things', $object);
+        $path = $this->getMethodAsPublic('getElementAccessPath')->invoke(
```
Comment (VincentLanglet) [inline tests/Admin/AdminHelperTest.php:87]: I would say that since the method is now private, there is no need to test it (instead the test should be done by the method which call this private method, `appendFormFieldElement`).

But since there is currently no `appendFormFieldElement` tests, I think this is ok for now.

**Comment by VincentLanglet** [comment]
```
1) Sonata\AdminBundle\Tests\Admin\Extension\LockExtensionTest::testPreUpdateIfAdminHasNoRequest
Method getModelManager may not return value of type NULL, its return declaration is ": Sonata\AdminBundle\Model\ModelManagerInterface"

2) Sonata\AdminBundle\Tests\Admin\Extension\LockExtensionTest::testPreUpdateIfObjectIsNotVersioned
Method getModelManager may not return value of type NULL, its return declaration is ": Sonata\AdminBundle\Model\ModelManagerInterface"

3) Sonata\AdminBundle\Tests\Admin\Extension\LockExtensionTest::testPreUpdateIfRequestDoesNotHaveLockVersion
Method getModelManager may not return value of type NULL, its return declaration is ": Sonata\AdminBundle\Model\ModelManagerInterface"

4) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditActionWithPreview
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

5) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithPreview
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

6) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtractEmpty
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

7) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtract
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

8) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtractCallsBreadcrumbs
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

9) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtractSetsSubject
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.
```

The warning are easy to fix.

1) Change the mock of FieldDescriptionCollection by, `new FieldDescriptionCollection()`.

2) Mock the return value of `getModelManager` in `LockExtensionTest`, for this I would recommend to change 
```
    private function configureAdmin(
        string $uniqid = '',
        ?Request $request = null,
        ?ModelManagerInterface $modelManager = null
    ): void {
```
to
```
    private function configureAdmin(
        ModelManagerInterface $modelManager
        string $uniqid = '',
        ?Request $request = null,
    ): void {
```
in order to always set a ModelManager

File: tests/Admin/AdminTest.php (hunk 20 above)
Comment (tambait) [reply tests/Admin/AdminTest.php:748]: I tried to do something, but not sure if I get the idea correctly. Also, LogicException messages could probably be more meaningful, I wasn't sure what to put so to be "generic" for now.

File: src/Admin/AbstractAdmin.php [hunk 22]
```diff
@@ -1882,8 +1895,12 @@ public function setConfigurationPool(Pool $configurationPool): void
         $this->configurationPool = $configurationPool;
     }
 
-    public function getConfigurationPool(): ?Pool
+    public function getConfigurationPool(): Pool
     {
+        if (null === $this->configurationPool) {
+            throw new \LogicException(sprintf('Pool has not been set.'));
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:1901]:
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no pool.',
                static::class
            ));
```

File: src/Admin/AbstractAdmin.php [hunk 23]
```diff
@@ -1892,8 +1909,12 @@ public function setRouteGenerator(RouteGeneratorInterface $routeGenerator): void
         $this->routeGenerator = $routeGenerator;
     }
 
-    public function getRouteGenerator(): ?RouteGeneratorInterface
+    public function getRouteGenerator(): RouteGeneratorInterface
     {
+        if (null === $this->routeGenerator) {
+            throw new \LogicException(sprintf('Router has not been set.'));
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:1915]:
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no route generator.',
                static::class
            ));
```

File: src/Admin/AbstractAdmin.php [hunk 24]
```diff
@@ -1911,12 +1932,16 @@ public function getBaseCodeRoute(): string
         return $this->getCode();
     }
 
-    public function getModelManager(): ?ModelManagerInterface
+    public function getModelManager(): ModelManagerInterface
     {
+        if (null === $this->modelManager) {
+            throw new \LogicException(sprintf('Manager has not been set.'));
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:1938]:
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no model manager.',
                static::class
            ));
```

File: src/Admin/AbstractAdmin.php [hunk 25]
```diff
@@ -2087,8 +2120,12 @@ public function setLabelTranslatorStrategy(LabelTranslatorStrategyInterface $lab
         $this->labelTranslatorStrategy = $labelTranslatorStrategy;
     }
 
-    public function getLabelTranslatorStrategy(): ?LabelTranslatorStrategyInterface
+    public function getLabelTranslatorStrategy(): LabelTranslatorStrategyInterface
     {
+        if (null === $this->labelTranslatorStrategy) {
+            throw new \LogicException(sprintf('LabelTranslatorStrategyInterface has not been set.'));
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2126]:
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no label translator strategy.',
                static::class
            ));
```

**Review by VincentLanglet — COMMENTED** [review]
You got the idea.

We already have message like
```
throw new \LogicException(sprintf(
                'Admin "%s" has no parent field description.',
                static::class
            ));
```

So I would go with something similar at beginning, until someone eventually improve the message to explain how to register/set all these services.

File: tests/Admin/AdminTest.php (hunk 20 above)
Comment (VincentLanglet) [reply tests/Admin/AdminTest.php:748]: You got the idea correctly. I juste made a minor comment on the exception message for consistency with others LogicException until we got something better.

**Comment by tambait** [comment]
> 1. Change the mock of FieldDescriptionCollection by, `new FieldDescriptionCollection()`.

Thanks. I'm not sure where to make this change, though maybe I was searching for/looking at wrong thing. When I searched through test files all `FieldDescriptionCollection` references were instantiated with `new FieldDescriptionCollection()` nor I found something like `CreateMock` etc

> 2. Mock the return value of `getModelManager` in `LockExtensionTest`, for this I would recommend to change

I've changed `configureAdmin` method but this caused errors in three other methods from the test class which didn't pass any parameters to original `configureAdmin` method or they passed just two parameters. I updated all of these to have `$this->configureAdmin('', null, $this->modelManager)` which removed the error, but not sure if this could have a unwanted impact on the test.

I also don't have idea how to solve phpStan errors :)

File: tests/Admin/Extension/LockExtensionTest.php [hunk 26]
```diff
@@ -235,7 +235,7 @@ private function configureFormMapper(): FormMapper
     private function configureAdmin(
         string $uniqid = '',
         ?Request $request = null,
-        ?ModelManagerInterface $modelManager = null
+        ModelManagerInterface $modelManager
```
Comment (VincentLanglet) [inline tests/Admin/Extension/LockExtensionTest.php:238]: Can you change the order of the param to 
`$modelManager, $uniqId, $request`.

This allow `$this->configureAdmin($this->modelManager)` instead of `$this->configureAdmin('', null, $this->modelManager)`

File: src/Admin/AbstractAdmin.php [hunk 27]
```diff
@@ -869,6 +868,9 @@ public function getBaseRouteName(): string
         return $this->cachedBaseRouteName;
     }
 
+    /**
+     * @phpstan-return class-string
+     */
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:873]: Remove this

**Review by VincentLanglet — COMMENTED** [review]
Edit: 

For one of your issue with Phpstan, there is some conflict between 
```
   /**
     * Returns list of supported sub classes.
     *
     * @return array<string, string>
     *
     * @phpstan-return array<string, class-string>
     */
    public function getSubClasses(): array;

    /**
     * Sets the list of supported sub classes.
     *
     * @param array<string, string> $subClasses
     *
     * @phpstan-param array<string, class-string> $subClasses
     */
    public function setSubClasses(array $subClasses): void;
```
And
```
/**
     * Returns subjectClass/class/subclass name managed
     * - subclass name if subclass parameter is defined
     * - subject class name if subject is defined
     * - class name if not.
     *
     * @phpstan-return class-string<T>
     */
    public function getClass(): string;
```

I think `<T>` should be removed for `getClass()`.

And then, you have to update
```
/**
     * Gets the subclass corresponding to the given name.
     *
     * @phpstan-return class-string<T>
     */
    protected function getSubClass(string $name): string
```
to remove the `<T>` too.

**Comment by VincentLanglet** [comment]
I would say you have to do
```
/**
     * The subclasses supported by the admin class.
     *
     * @var string[]
     *
     * @phpstan-var array<string, class-string<T>>
     */
    private $subClasses = [];
```
In AbstractAdmin

Then
```
/**
     * Returns list of supported sub classes.
     *
     * @return array<string, string>
     *
     * @phpstan-return array<string, class-string<T>>
     */
    public function getSubClasses(): array;

    /**
     * Sets the list of supported sub classes.
     *
     * @param array<string, string> $subClasses
     *
     * @phpstan-param array<string, class-string<T>> $subClasses
     */
    public function setSubClasses(array $subClasses): void;

    /**
     * Returns true if the admin has the sub classes.
     *
     * @phpstan-param class-string<T> $name
     */
    public function hasSubClass(string $name): bool;
```
In the AdminInterface

File: src/Admin/AbstractAdmin.php [hunk 28]
```diff
@@ -2676,9 +2725,10 @@ protected function configureDefaultFilterValues(array &$filterValues): void
      *   $sortValues['_sort_by'] = 'foo'
      *   $sortValues['_sort_order'] = 'DESC'
      *
+     * @param array<string, string|int>
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2728]:
```suggestion
     * @param array<string, string|int> $sortValues
     *
```

**Comment by VincentLanglet** [comment]
For the phpstan error inside attacheInlineValidator, I would recommend to write
```
// @phpstan-ignore-next-line
$admin->validate($errorElement, $object);
```
It should be removed: https://github.com/sonata-project/SonataAdminBundle/issues/6233

**Comment by VincentLanglet** [comment]
Do you encounter some issues or is it just a question of time ?
I don't want to hurry at all.

I just wanted to tell you that if you encounter issues and need help, don't hesitate to ping me. ;)
I'm working on multiple PR and don't always think to take a look at yours.

**Comment by tambait** [comment]
> Do you encounter some issues or is it just a question of time ?
> I don't want to hurry at all.
> 
> I just wanted to tell you that if you encounter issues and need help, don't hesitate to ping me. ;)
> I'm working on multiple PR and don't always think to take a look at yours.

Thanks. Little bit of both :) I understand that you're working on multiple PR's so I was looking to solve issues that remained by myself. If I didn't oversee something think that "major" problems are with FieldDescriptionCollection in [tests](https://github.com/sonata-project/SonataAdminBundle/pull/6539#issuecomment-717432398) as well as phpStan conflict that you commented [here](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831).

**Comment by VincentLanglet** [comment]
> Thanks. Little bit of both :) I understand that you're working on multiple PR's so I was looking to solve issues that remained by myself. If I didn't oversee something think that "major" problems are with FieldDescriptionCollection in [tests](https://github.com/sonata-project/SonataAdminBundle/pull/6539#issuecomment-717432398) as well as phpStan conflict that you commented [here](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831).

Great.

I think I gave everything needed for the phpstan issue.

For the fieldDescriptionCollection, it's weird indeed. I saw no `createMock()` calls.
But for instance, `testCreateActionWithPreview` will execute this code in the createAction
```
elseif ($this->isPreviewRequested()) {
                // pick the preview template if the form was valid and preview was requested
                $templateKey = 'preview';
                $this->admin->getShow();
            }
```

Since the `admin` is a mock in the tests (see the setup), Phpunit automatically create a mock for the return value.
Before the type was `?FieldDescriptionCollection` so it was returning null.
Now you restrict to `FieldDescriptionCollection` so it will mock FieldDescriptionCollection.

You should then, in the setup of CRUDControllerTest, add 
```
$this->admin->method('getShow')->willReturn(new FieldDescriptionCollection())
```

Same for others issues (maybe ?)

**Comment by tambait** [comment]
> I think I gave everything needed for the phpstan issue.

Yes you did. All phpStan fixes are implemented except, I think, the one you [asked](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831) @core23 about

It's the error on line 39 in `FieldDescriptionInterface.php` which states `PHPDoc tag @return with type string|null is not subtype of native type string.`

>Since the admin is a mock in the tests (see the setup), Phpunit automatically create a mock for the return value.
Before the type was ?FieldDescriptionCollection so it was returning null.
Now you restrict to FieldDescriptionCollection so it will mock FieldDescriptionCollection.`

> You should then, in the setup of CRUDControllerTest, add
`$this->admin->method('getShow')->willReturn(new FieldDescriptionCollection())`


Great. I was trying to do something similar with `willReturn` but in a wrong way. For the change you suggested, if I put the code inside setUp method it triggers two failures `Failed asserting that two variables reference the same object` in

> 1) CRUDControllerTest::testHistoryViewRevisionAction :2969
> 2) CRUDControllerTest::testHistoryCompareRevisionsActionAction :3184

however if I put the the line inside, for example, `testEditActionWithPreview` method it goes without failures. I guess I could do the same for all methods which produce warning in the test, but I'm not sure if placing the line in setUp method and fixing new failures could produce same results without additional side effects.

**Comment by VincentLanglet** [comment]
> > I think I gave everything needed for the phpstan issue.
> 
> Yes you did. All phpStan fixes are implemented except, I think, the one you [asked](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831) @core23 about
> 
> It's the error on line 39 in `FieldDescriptionInterface.php` which states `PHPDoc tag @return with type string|null is not subtype of native type string.`

You can follow: https://github.com/sonata-project/SonataAdminBundle/pull/6539#issuecomment-717472455

> > Since the admin is a mock in the tests (see the setup), Phpunit automatically create a mock for the return value.
> > Before the type was ?FieldDescriptionCollection so it was returning null.
> > Now you restrict to FieldDescriptionCollection so it will mock FieldDescriptionCollection.`
> 
> > You should then, in the setup of CRUDControllerTest, add
> > `$this->admin->method('getShow')->willReturn(new FieldDescriptionCollection())`
> 
> Great. I was trying to do something similar with `willReturn` but in a wrong way. For the change you suggested, if I put the code inside setUp method it triggers two failures `Failed asserting that two variables reference the same object` in
> 
> > 1. CRUDControllerTest::testHistoryViewRevisionAction :2969
> > 2. CRUDControllerTest::testHistoryCompareRevisionsActionAction :3184
> 
> however if I put the the line inside, for example, `testEditActionWithPreview` method it goes without failures. I guess I could do the same for all methods which produce warning in the test, but I'm not sure if placing the line in `setUp` method and fixing new failures could produce same results without additional side effects.

It's because the method is already mocked.
I would say then that it's easier to put the line inside `testEditActionWithPreview` instead of `setUp`.

**Comment by tambait** [comment]
@VincentLanglet 

PHPUnit Tests are hopefully good now, but I'm still puzzled with 

>  Line 39   FieldDescriptionInterface.php                                                   
>   PHPDoc tag @return with type string|null is not subtype of native type string.  

If I run phpstan locally on my branch it goes without error. If locally run phpstan on master branch I get same error check as above. Master branch has

```
    /**
     * NEXT_MAJOR: Restrict return type to string.
     *
     * Returns the name, the name can be used as a form label or table header.
     *
     * @return string|null the name
     */
    public function getName(): string;
```
while in my branch I have

```
    /**
     * Returns the name, the name can be used as a form label or table header.
     *
     * @return string the name
     */
    public function getName(): string;
```
When I look at annotation about the error, I have feeling like it's displaying code from my branch but complaining about the code from the master branch :smiley: 

There are few errors reported by psalm

```
Error: Class or interface Symfony\Bundle\MakerBundle\Maker\AbstractMaker does not exist
Error: Class or interface Symfony\Component\Yaml\Yaml does not exist
Error: Sonata\AdminBundle\Maker\AdminMaker depends 
on class or interface symfony\bundle\makerbundle\maker\abstractmaker that does not exist

```
and there was also
```
$this->securityInformation with declared type 'array<string, array<array-key, string>>' 
cannot be assigned type 'array<string, array<array-key, string>|string>'
```
where I changed declaration for `$securityInformation` from 
`@var array<string, string[]> 'role' => ['permission', 'permission']`

to 

`@var array<string, string[]|string> 'role' => ['permission', 'permission']`

**Comment by VincentLanglet** [comment]
I merged 3.x into Master

Wait for https://github.com/sonata-project/SonataAdminBundle/pull/6557, maybe it will fix your issues.

**Comment by tambait** [comment]
It could be that something was messed during resolving conflicts with earlier version of master. I now modified phpdoc block by removing `@return string the name`, resolved the conflict with the  (updated) master and phpstan error disappeared
```
    /**
     * Returns the name, the name can be used as a form label or table header.
     *
     * @return string the name
     */
     public function getName(): string
```

**Review by VincentLanglet — DISMISSED** [review]

**Comment by VincentLanglet** [comment]
Please review @sonata-project/contributors

**Comment by VincentLanglet** [comment]
@tambait Do you have some time to rebase ? Then I'll merge it.

**Comment by tambait** [comment]
> @tambait Do you have some time to rebase ? Then I'll merge it.

Yes, thank for notifying me (I missed notification about conflicts). 
After this one is merged, is there some "to do" list or something similar I can look at for future work/contributions? At least for me personally, list you've made for param/return type hints was very useful to get involved with the project

File: src/Admin/Pool.php [hunk 29]
```diff
@@ -63,32 +57,28 @@ class Pool
...
      */
     protected $propertyAccessor;
 
     /**
-     * NEXT_MAJOR: change to TemplateRegistryInterface.
-     *
-     * @var MutableTemplateRegistryInterface
+     * @param array<string, mixed> $options
      */
-    private $templateRegistry;
```
Comment (VincentLanglet) [inline src/Admin/Pool.php:80]: Seems like this should not be removed

**Comment by VincentLanglet** [comment]
You messed up the phpdoc with your fix up

File: src/Admin/Pool.php (hunk 29 above)
Comment (tambait) [reply src/Admin/Pool.php:80]: Yes, thanks. Not sure if I messed up something. Should I also replace MutableTemplateRegistryInterface with TemplateRegistryInterfac?  Is that comment is for 4.0 version (not sure I saw it before)

**Comment by tambait** [comment]
> You messed up the phpdoc with your fix up

Oh. I just run php-cs-fixer. How can I fix that?

**Comment by VincentLanglet** [comment]
It should be
```
/**
  * NEXT_MAJOR: change to TemplateRegistryInterface.
  *
  * @var MutableTemplateRegistryInterface
  */
private $templateRegistry;
```
because
```
     * @param array<string, mixed> $options
```
is related to the constructor.

**Comment by tambait** [comment]
> It should be
> 
> ```
> /**
>   * NEXT_MAJOR: change to TemplateRegistryInterface.
>   *
>   * @var MutableTemplateRegistryInterface
>   */
> private $templateRegistry;
> ```
> 
> because
> 
> ```
>      * @param array<string, mixed> $options
> ```
> 
> is related to the constructor.

Thanks. Just realized what was problem

**Review by VincentLanglet — APPROVED** [review]

**Review by VincentLanglet — COMMENTED** [review]
I think this https://github.com/sonata-project/SonataAdminBundle/pull/6539/files#diff-0850515af24b81ce716f53b8eb0b71fae91dcbdcc88b28d95a83886504e1c8f4R329

Should be Pool|null

**Comment by VincentLanglet** [comment]
Anyway I'll work on the `null` issue in https://github.com/sonata-project/SonataAdminBundle/pull/6614

**Comment by VincentLanglet** [comment]
Thanks @tambait.

If you want to continue helping us, I would recommend the milestone 4.0
https://github.com/sonata-project/SonataAdminBundle/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0

Don't hesitate to ask anything before starting an issue. I already start working on some.

**Comment by tambait** [comment]
> Thanks @tambait.
> 
> If you want to continue helping us, I would recommend the milestone 4.0
> https://github.com/sonata-project/SonataAdminBundle/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0
> 
> Don't hesitate to ask anything before starting an issue. I already start working on some.

Great thanks, I'll take a look.
